ISSUE 2849: '~' character in object name causes V4 SignatureDoesNotMatch

tpodowd opened the pull request:
When calculating the canonical request for signing with V4 signatures,
the following characters should not be url encoded in the Canonical URI:
 A-Z, a-z, 0-9, -, ., _, ~, /

All characters were handled correctly except ~.

tpodowd commented:
I also added a simple test. There are not many V4 tests right now anyway.

    $ nosetests unit/s3/test_connection.py
    ............
    ----------------------------------------------------------------------
    Ran 12 tests in 0.018s
    OK

pslawski commented:
It seems like our efforts were duplicated. The following pull request also resolves this issue, #2844, as well #2846.

tpodowd commented:
Hi @pslawski, I checked out your branch and tried it out manually. Looks good to me and yes, your branch also fixes the ~ issue. I'd like to close this pull request so that yours gets in. Can you take a look at my unit test and see if it is something you would like to put in your branch? I didn't look at what tests you had. Let me know and I'll close this.

tpodowd commented:
@jamesls - original issue is now fixed in latest develop so I rebased this pull request to remove the identical fix and instead to just keep the test case for the original issue. Hopefully an easy merge decision.

    [tpodowd@TMPro s3 (issue2849<>)]$ nosetests-2.7 test_connection.py
    ...
    ----------------------------------------------------------------------
    Ran 3 tests in 42.875s
    
    OK
